### PR TITLE
feature(distinctUntilSomeChanged)

### PR DESCRIPTION
**Pull request by hoebbelsB**
implemented `distinctUntilSomeChanged` operator. Added some docs and implemented tests (based off of `distinctUntilKeyChanged` from rxjs.

### CODE REVIEW

File: libs/state/spec/core/operators/distinctUntilKeysChanged.spec.ts [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+      expectSubscriptions(e1.subscriptions).toBe(e1subs);
+    });
+  });
+
+  it('should emit if source emits single element only', () => {
+    testScheduler.run(({ cold, expectObservable, expectSubscriptions }) => {
+      const values = { a: { val: 1 } };
+      const e1 = cold('--a--|', values);
+      const e1subs = '^----!';
+      const expected = '--a--|';
```
Comment (BioPhoton) [inline libs/state/spec/core/operators/distinctUntilKeysChanged.spec.ts:239]:
```suggestion
      const e1 =    cold('--a--|', values);
      const e1subs =    '^----!';
      const expected = '--a--|';
```

File: libs/state/spec/core/operators/distinctUntilKeysChanged.spec.ts [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+  testScheduler = new TestScheduler(jestMatcher);
+});
+
+/** @test {distinctUntilSomeChanged} */
+describe('distinctUntilSomeChanged operator', () => {
+  it('should distinguish between values', () => {
+    testScheduler.run(({ cold, expectObservable, expectSubscriptions }) => {
+      const values = { a: { val: 1 }, b: { val: 2 } };
+      const e1 = cold('--a--a--a--b--b--a--|', values);
+      const e1subs = '^-------------------!';
```
Comment (BioPhoton) [inline libs/state/spec/core/operators/distinctUntilKeysChanged.spec.ts:34]:
```suggestion
      const e1subs =  '^-------------------!';
```

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+export type KeyCompareMap<T extends object> = {
+  [K in keyof T]: CompareFn<T[K]>;
+};
+
+/**
+ * @description
+ *
+ * Returns an Observable that emits all items emitted by the source Observable that are distinct by comparison from
+ * the previous item. You can provide a custom comparison for each key individually by setting a `KeyCompareMap<T>`.
+ * If no comparison is provided for a specified key, an equality check is used by default.
```
Comment (BioPhoton) [inline libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:20]: I don'T see the equality check happening in the code. 
Also, keys or keymap is a required param. 
IMHO this is wrong.

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ * If properties of the source change, which are not specified for comparison, no change will be emitted.
```
Comment (BioPhoton) [inline libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:22]: I did not see a test for that behavior

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+/**
+ * @internal
+ */
+export type KeyCompareMap<T extends object> = {
+  [K in keyof T]: CompareFn<T[K]>;
+};
+
+/**
+ * @description
+ *
```
Comment (BioPhoton) [inline libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:17]: I would add information why the naming was picked and a link to MDN

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+): MonoTypeOperatorFunction<T>;
+/**
+ * @internal
+ */
+export function distinctUntilSomeChanged<T extends object, K extends keyof T>(
+  keysOrMap: K[] | KeyCompareMap<T>,
+  compare?: CompareFn<T[K]>
+): MonoTypeOperatorFunction<T> {
+  let distinctCompare: CompareFn<T>;
+  const defaultCompare: CompareFn<T[K]> = (oldVal, newVal) => oldVal === newVal;
```
Comment (hoebbelsB) [inline libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:155]: move defaultCompare to top. maybe we find a way to memoize this without the need of any dependency?

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts (hunk 3 above)
Comment (hoebbelsB) [reply libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:20]: `const defaultCompare: CompareFn<T[K]> = (oldVal, newVal) => oldVal === newVal;` this is the equality check. So this is true :)

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts (hunk 4 above)
Comment (hoebbelsB) [reply libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:22]: see the test named `should ignore changes of other keys`

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts (hunk 3 above)
Comment (BioPhoton) [reply libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:20]: Ah! 
> For a key

Got it!

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts (hunk 5 above)
Comment (hoebbelsB) [reply libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:17]: agree

**Comment by BioPhoton** [comment]
We also need to update the docs with this operator

File: libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts (hunk 5 above)
Comment (hoebbelsB) [reply libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts:51]: done

**Comment by hoebbelsB** [comment]
> We also need to update the docs with this operator

i have added a [snippet](https://github.com/BioPhoton/rx-angular/pull/81/files#diff-9ec871aba9ba6a1adb710d564547acb7) . #84

**Pull request merged**